### PR TITLE
feat(slider): Add value label to discrete variant

### LINKED ISSUE #2983: Indicator for new discrete slider
Follows #2982
- [ ] `mdc-slider__indicator` and `mdc-slider__indicator-value`
- [ ] ability to update indicator shape based on indicator value (see #1426 and #1437)

### PR DESCRIPTION
**Pull request by EstebanG23**
Fixes #2983

### CODE REVIEW

File: packages/mdc-slider/README.md [hunk 1]
```diff
@@ -201,6 +225,9 @@ instance that was affected.
 | `addClass(className: string) => void` | Adds a class `className` to the root element |
 | `removeClass(className: string) => void` | Removes a class `className` from the root element |
 | `setThumbAttribute(name: string, value: string) => void` | Sets an attribute `name` to the value `value` on the thumb element. |
+| `setValueLabelPath(value: string) => void` | Sets the path of the value label Element to the value `value`. |
+| `setValueLabelText(xValue: string, text: string) => void` | Sets the inner text and translateX to the assigned values. |
+| `removeValueLabelTextStyle() => void` | Removes the style attribute for the value label text Element. |
```
Comment (bonniezhou) [inline packages/mdc-slider/README.md]: Missing a few methods: `getDigitWidth`, `getCommaWidth`, `registerThumbEventHandler`, `deregisterThumbEventHandler`

File: packages/mdc-slider/_mixins.scss [hunk 2]
```diff
@@ -48,3 +48,15 @@
   }
 }
 
+@mixin mdc-slider-value-label-color($color) {
```
Comment (bonniezhou) [inline packages/mdc-slider/_mixins.scss]: The default should be that changing the value label color also updates the thumb to the same color, so `@include` the `mdc-slider-thumb-color` mixin inside this one

File: packages/mdc-slider/foundation.js [hunk 3]
```diff
@@ -414,6 +471,237 @@ class MDCSliderFoundation extends MDCFoundation {
...
+    }
+    if (this.adapter_.isRTL()) {
+      xValue = xValue + localeStringWidth;
+    }
+
+    // Calculates any extra translate value on either side if the slider is near the ends
+    if (this.value_.toString().length > 2) {
+      topLobeHorizontal = localeStringWidth - (2 * digitWidth);
+    }
+    const extra = topLobeHorizontal - 30;
```
Comment (bonniezhou) [inline packages/mdc-slider/foundation.js]: Update these constants to use the new names `extraHorizontalWidth`, `extraHorizontalWidthLeft`, `extraHorizontalWidthRight`

File: packages/mdc-slider/foundation.js [hunk 4]
```diff
@@ -414,6 +471,237 @@ class MDCSliderFoundation extends MDCFoundation {
...
+   * @return {number}
+   */
+  calcValueLabelTextXValue_(translatePx) {
+    const localeStringWidth = this.calcLocaleStringWidth_();
+    let extraTranslateValue = 0;
+    let topLobeHorizontal = 0;
+    const digitWidth = this.adapter_.getDigitWidth();
+
+    // Calculates default translateX value for the size of the text
+    let xValue = (34 - localeStringWidth);
```
Comment (bonniezhou) [inline packages/mdc-slider/foundation.js]: What's 34?

File: packages/mdc-slider/foundation.js [hunk 5]
```diff
@@ -233,13 +280,18 @@ class MDCSliderFoundation extends MDCFoundation {
     if (this.inTransit_ && this.adapter_.eventTargetHasClass(evt.target, cssClasses.TRACK_FILL)) {
       this.setInTransit_(false);
     }
+    if (this.isDiscrete_ && this.adapter_.eventTargetHasClass(evt.target, 'mdc-slider__value-label-text')
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply packages/mdc-slider/foundation.js]: 👍

File: test/unit/mdc-slider/foundation.test.js [hunk 6]
```diff
@@ -79,6 +81,7 @@ test('#init registers all necessary event handlers for the component', () => {
   td.verify(mockAdapter.registerEventHandler('keydown', isA(Function)));
   td.verify(mockAdapter.registerEventHandler('keyup', isA(Function)));
   td.verify(mockAdapter.registerEventHandler('transitionend', isA(Function)));
+  td.verify(mockAdapter.registerThumbEventHandler('blur', isA(Function)));
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply test/unit/mdc-slider/foundation.test.js]: Got it!

File: packages/mdc-slider/foundation.js [hunk 7]
```diff
@@ -450,6 +745,15 @@ class MDCSliderFoundation extends MDCFoundation {
     this.toggleClass_(cssClasses.IN_TRANSIT, this.inTransit_);
   }
 
+  /**
+   * Toggles the pressed state of the slider
+   * @param {boolean} pressed
+   */
+  setPressed_(pressed) {
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply packages/mdc-slider/foundation.js]: Done!

File: packages/mdc-slider/foundation.js [hunk 8]
```diff
@@ -296,8 +348,13 @@ class MDCSliderFoundation extends MDCFoundation {
       return;
     }
 
+    if (this.isDiscrete_) {
+      this.interactingWithSlider_ = true;
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply packages/mdc-slider/foundation.js]: True makes sense.

File: packages/mdc-slider/foundation.js [hunk 9]
```diff
@@ -414,6 +471,237 @@ class MDCSliderFoundation extends MDCFoundation {
...
+    // Value to add to the sides of the toplobe to adjust for big numbers
+    const extraHorizontalWidth = Math.max(labelHorizontalWidth - maxTopLobeHorizontal, 0);
+
+    // Distributes the extra length to left and right
+    let extraHorizontalWidthLeft = extraHorizontalWidth * 3 / 4;
+    let extraHorizontalWidthRight = extraHorizontalWidth / 4;
+
+    // If the thumb is reaching the ends of the slider then edit the left and right to make sure
+    // it does not bleed off the screen.
+    const distanceFromLeft = this.adapter_.isRTL() ? this.rect_.width - translatePx - 15 : translatePx - 15;
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply packages/mdc-slider/foundation.js]: Got it!

File: packages/mdc-slider/foundation.js [hunk 10]
```diff
@@ -414,6 +471,237 @@ class MDCSliderFoundation extends MDCFoundation {
...
+    const topNeckCornerCenterY = Math.sqrt(Math.pow(topLobeRadius+topNeckRadius, 2) -
+      Math.pow(15 - topLobeHorizontal/2, 2));
+    // Distance between the top lobe and the bottom lobe
+    const centersDifference = 40;
+    // Radius of the bottom lobe
+    const bottomLobeRadius = 6;
+    // Radius of the bottom neck
+    const bottomNeckRadius = 4.5;
+    // Angle of the bottom neck
+    const bottomNeckTheta = 5 * Math.PI / 18;
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply packages/mdc-slider/foundation.js]: 5pi/18 = 50degrees which is used as the angle for the bottom neck.

File: packages/mdc-slider/foundation.js [hunk 11]
```diff
@@ -414,6 +471,237 @@ class MDCSliderFoundation extends MDCFoundation {
...
+      y: topLobeRadius + centersDifference - bottomNeckHeight + offsetY,
+    };
+    const pointD = {
+      x: rightStemX + bottomNeckRadius - (Math.cos(bottomNeckTheta) * bottomNeckRadius) + 1 + offsetX,
+      y: (topLobeRadius + centersDifference - bottomNeckHeight) +
+         (Math.sin(bottomNeckTheta) * bottomNeckRadius) + offsetY,
+    };
+    const pointE = {
+      x: topLobeRadius + offsetX,
+      y: 62 + offsetY,
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply packages/mdc-slider/foundation.js]: Height of entire value label, updated!

File: packages/mdc-slider/foundation.js [hunk 12]
```diff
@@ -414,6 +471,237 @@ class MDCSliderFoundation extends MDCFoundation {
...
+      end.x = end.x - extraHorizontalWidthLeft;
+      pointA.x = pointA.x + extraHorizontalWidthRight;
+    }
+
+    // When the slider reaches a certain point close to the edges the extra on the side closest to the edge
+    // is negative and the angles of the neck need to be updates to create smooth value label
+    if (extraHorizontalWidthLeft < 0 && extraHorizontalWidth > 0) {
+      const topLobeHorizontalLeft = this.adapter_.isRTL() ? (this.rect_.width - translatePx) * 2 : translatePx * 2;
+
+      const leftTopNeckCornerTheta = Math.acos((15 - (topLobeHorizontalLeft)/2)/(topLobeRadius+topNeckRadius));
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply packages/mdc-slider/foundation.js]: Updated from prior comment!

File: test/unit/mdc-slider/foundation-pointer-events.test.js [hunk 13]
```diff
@@ -111,6 +111,152 @@ function createTestSuiteForPointerEvents(downEvt, moveEvt, upEvt, pageXObj = (pa
...
+    const {isA} = td.matchers;
+
+    td.when(mockAdapter.computeBoundingRect()).thenReturn({left: 0, width: 100});
+    td.when(mockAdapter.hasClass(cssClasses.DISCRETE)).thenReturn(true);
+    td.when(mockAdapter.getCommaWidth()).thenReturn(3);
+    td.when(mockAdapter.getDigitWidth()).thenReturn(9);
+    foundation.init();
+    raf.flush();
+
+    foundation.setMax(100000);
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply test/unit/mdc-slider/foundation-pointer-events.test.js]: The next line calls the interaction start handler and passes the pageX value given which is 100 in this case meaning its the max.

File: test/unit/mdc-slider/foundation.test.js [hunk 14]
```diff
@@ -610,3 +614,170 @@ test('#handleTransitionEnd no-op when inTransit_ is true and event target is not
 
   raf.restore();
 });
+
+test('#handleTransitionEnd sets pressed_ to true when all parameters are true', () => {
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply test/unit/mdc-slider/foundation.test.js]: 👍

File: test/unit/mdc-slider/foundation.test.js [hunk 15]
```diff
@@ -610,3 +614,170 @@ test('#handleTransitionEnd no-op when inTransit_ is true and event target is not
...
+  td.when(mockAdapter.eventTargetHasClass(mockEvent.target, 'mdc-slider__value-label-text')).thenReturn(true);
+  foundation.handleTransitionEnd(mockEvent);
+  raf.flush();
+
+  td.verify(mockAdapter.addClass(cssClasses.PRESSED));
+
+  raf.restore();
+});
+
+test('#handleTransitionEnd no-op when isDiscrete_ is false', () => {
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply test/unit/mdc-slider/foundation.test.js]: Done!

File: test/unit/mdc-slider/foundation.test.js [hunk 16]
```diff
@@ -610,3 +614,170 @@ test('#handleTransitionEnd no-op when inTransit_ is true and event target is not
...
+  td.when(mockAdapter.eventTargetHasClass(mockEvent.target, 'mdc-slider__value-label')).thenReturn(true);
+  foundation.handleTransitionEnd(mockEvent);
+  raf.flush();
+
+  td.verify(mockAdapter.addClass(cssClasses.PRESSED), {times: 0});
+
+  raf.restore();
+});
+
+test('#handleTransitionEnd no-op for pressed_ when active_ is false', () => {
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply test/unit/mdc-slider/foundation.test.js]: 👍

File: test/unit/mdc-slider/foundation.test.js [hunk 17]
```diff
@@ -610,3 +614,170 @@ test('#handleTransitionEnd no-op when inTransit_ is true and event target is not
...
+  const mockKeyboardEvent = {
+    preventDefault: () => {},
+    keyCode: 37,
+  };
+  foundation.handleKeydown(mockKeyboardEvent);
+  raf.flush();
+
+  const mockEvent = {
+    target: {
+      classList: ['mdc-slider__value-label'],
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply test/unit/mdc-slider/foundation.test.js]: You're right, updated! And I used the cssClasses

File: test/unit/mdc-slider/foundation.test.js [hunk 18]
```diff
@@ -610,3 +614,170 @@ test('#handleTransitionEnd no-op when inTransit_ is true and event target is not
...
+  td.when(mockAdapter.eventTargetHasClass(mockEvent.target, 'mdc-slider__value-label')).thenReturn(true);
+  foundation.handleTransitionEnd(mockEvent);
+  raf.flush();
+
+  td.verify(mockAdapter.addClass(cssClasses.PRESSED), {times: 0});
+
+  raf.restore();
+});
+
+test('#handleTransitionEnd no-op for pressed_ when event target is not mdc-slider__value-label', () => {
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply test/unit/mdc-slider/foundation.test.js]: 👍

File: test/unit/mdc-slider/foundation.test.js [hunk 19]
```diff
@@ -610,3 +614,170 @@ test('#handleTransitionEnd no-op when inTransit_ is true and event target is not
...
+test('#handleThumbBlur no-op when interactingWithSlider is true', () => {
+  const {foundation, mockAdapter} = setupTest();
+  const raf = createMockRaf();
+  td.when(mockAdapter.computeBoundingRect()).thenReturn({left: 0, width: 0});
+  td.when(mockAdapter.hasClass(cssClasses.DISCRETE)).thenReturn(false);
+  foundation.init();
+  raf.flush();
+
+  // To set this.interactingWithSlider_ to true
+  const mockKeyboardEvent = {
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply test/unit/mdc-slider/foundation.test.js]: Done!

File: packages/mdc-slider/README.md (hunk 1 above)
Comment (EstebanG23) [reply packages/mdc-slider/README.md]: Done!

File: packages/mdc-slider/_mixins.scss (hunk 2 above)
Comment (EstebanG23) [reply packages/mdc-slider/_mixins.scss]: Done!

File: packages/mdc-slider/foundation.js (hunk 3 above)
Comment (EstebanG23) [reply packages/mdc-slider/foundation.js]: Got it!

File: packages/mdc-slider/foundation.js (hunk 4 above)
Comment (EstebanG23) [reply packages/mdc-slider/foundation.js]: It is the width of the svg workspace, i added a constant and description!

File: packages/mdc-slider/README.md [hunk 20]
```diff
@@ -201,10 +225,18 @@ instance that was affected.
 | `addClass(className: string) => void` | Adds a class `className` to the root element |
 | `removeClass(className: string) => void` | Removes a class `className` from the root element |
 | `setThumbAttribute(name: string, value: string) => void` | Sets an attribute `name` to the value `value` on the thumb element. |
+| `setValueLabelPath(value: string) => void` | Sets the path of the value label Element to the value `value`. |
+| `setValueLabelText(text: string) => void` | Sets the inner text and translateX to the assigned values. |
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply packages/mdc-slider/README.md]: Fixed it!

File: packages/mdc-slider/index.js [hunk 21]
```diff
@@ -118,14 +134,36 @@ class MDCSlider extends MDCComponent {
         addClass: (className) => this.root_.classList.add(className),
         removeClass: (className) => this.root_.classList.remove(className),
         setThumbAttribute: (name, value) => this.thumb_.setAttribute(name, value),
+        setValueLabelPath: (value) => this.valueLabel_.setAttribute('d', value),
+        setValueLabelText: (text) => {
+          this.valueLabelText_.textContent = text;
+        },
+        setValueLabelStyleProperty: (propertyName, value) => {
+          this.valueLabelText_.setAttribute(propertyName, value);
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply packages/mdc-slider/index.js]: Done!

File: packages/mdc-slider/index.js [hunk 22]
```diff
@@ -118,14 +134,36 @@ class MDCSlider extends MDCComponent {
         addClass: (className) => this.root_.classList.add(className),
         removeClass: (className) => this.root_.classList.remove(className),
         setThumbAttribute: (name, value) => this.thumb_.setAttribute(name, value),
+        setValueLabelPath: (value) => this.valueLabel_.setAttribute('d', value),
+        setValueLabelText: (text) => {
+          this.valueLabelText_.textContent = text;
+        },
+        setValueLabelStyleProperty: (propertyName, value) => {
```
[the remark this answers is not in the document]
Comment (EstebanG23) [reply packages/mdc-slider/index.js]: Got it!

**Pull request merged**